File: abbott-mobile-mylan/platforms/android/AbbottMobile/src/com/qapint/app/models/CallReport.java
package com.qapint.app.models;

import java.text.ParseException;
import java.util.Date;
import com.qapint.app.utils.DateUtils;
import org.json.JSONException;
import org.json.JSONObject;

public class CallReport extends Entity{
    public static final String DATE_FIELD_NAME = "Date_Time__c";
    public static final String TYPE_FIELD_NAME = "Type__c";
    public static final String ORGANISATION_FIELD_NAME = "Organisation__r";
    public static final String ORGANISATION_NAME_FIELD_NAME = "Name";
    public static final String ORGANISATION_CITY_FIELD_NAME = "BillingCity";
    public static final String ORGANISATION_STREET_FIELD_NAME = "BillingStreet";

    private Date dateOfVisit;
    private String organizationName;
    private String organizationCity;
    private String organizationAddress;

    private void parseFromJSON(JSONObject object) throws JSONException, ParseException{
        String date = object.getString(DATE_FIELD_NAME);
        dateOfVisit = DateUtils.SFGMTToLocal(date);
        if(object.has(ORGANISATION_FIELD_NAME)){
            JSONObject organisation = object.getJSONObject(ORGANISATION_FIELD_NAME);
            organizationName = organisation.getString(ORGANISATION_NAME_FIELD_NAME);
            organizationCity = organisation.getString(ORGANISATION_CITY_FIELD_NAME);
            organizationAddress = organisation.getString(ORGANISATION_STREET_FIELD_NAME);
        }else{
            organizationName = object.getString(String.format("%s.%s", ORGANISATION_FIELD_NAME, ORGANISATION_NAME_FIELD_NAME));
            organizationCity = object.getString(String.format("%s.%s", ORGANISATION_FIELD_NAME, ORGANISATION_CITY_FIELD_NAME));
            organizationAddress = object.getString(String.format("%s.%s", ORGANISATION_FIELD_NAME, ORGANISATION_STREET_FIELD_NAME));
        }
    }

    public CallReport(JSONObject object) throws JSONException, ParseException {
        super(object);
        parseFromJSON(object);
    }

    public Date getDateOfVisit(){
        return dateOfVisit;
    }

    public String getOrganization(){
        return String.format("%s\n%s %s", organizationName, organizationAddress, organizationCity);
    }

    public String getDateOfVisitFormatted() {
        return DateUtils.toTraditional(getDateOfVisit());
    }
}

File: abbott-mobile-mylan/platforms/ios/AbbottMobile/AbbottMobile/Classes/DAL/CallReport.m
//
//  CallReport.m
//  AbbottMobile
//
//  Created by Alexander Voronov on 12/6/14.
//  Copyright (c) 2014 Qap, Inc. All rights reserved.
//

#import "CallReport.h"
#import "DateUtils.h"

@implementation CallReport

@synthesize identifier;
@synthesize dateOfVisit;
@synthesize type;
@synthesize duration;
@synthesize organizationName;
@synthesize organizationAddress;
@synthesize organizationCity;

+ (NSString *)tableName
{
    return @"CallReport";
}

+ (NSString *)idFieldName
{
    return @"Id";
}

+ (NSString *)dateFieldName
{
    return @"Date_Time__c";
}

+ (NSString *)typeFieldName
{
    return @"Type__c";
}

+ (NSString *)durationFieldName
{
    return @"Duration__c";
}

+ (NSString *)organisationFieldName
{
    return @"Organisation__r";
}

+ (NSString *)organisationNameFieldName
{
    return @"Name";
}

+ (NSString *)organisationCityFieldName
{
    return @"BillingCity";
}

+ (NSString *)organisationStreetFieldName
{
    return @"BillingStreet";
}

- (NSString *)fullAddress
{
    return [NSString stringWithFormat:@"%@ %@", self.organizationAddress, self.organizationCity];
}

#pragma mark - 

- (void)dealloc
{
    self.identifier = nil;
    self.dateOfVisit = nil;
    self.type = nil;
    self.duration = nil;
    self.organizationName = nil;
    self.organizationAddress = nil;
    self.organizationCity = nil;
    
    [super dealloc];
}

@end
